File: fbot_manipulator/src/utils/save_arm_pose.cpp
#include <chrono>
#include <fstream>
#include <iostream>
#include <regex>
#include <sstream>
#include <string>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/joint_state.hpp>

using sensor_msgs::msg::JointState;

class SaveArmPose : public rclcpp::Node
{
public:
    SaveArmPose()
    : Node("save_arm_pose")
    {
        this->declare_parameter<std::string>(
            "xacro_path",
            "/home/insider/fbot_ws/src/xarm_ros2/xarm_moveit_config/srdf/_xarm6_macro.srdf.xacro");
            //TODO: Change that value in robot
        xacro_path_ = this->get_parameter("xacro_path").as_string();

        group_name_ = "${prefix}xarm6";
        default_arm_joint_order_ = {"joint1", "joint2", "joint3", "joint4", "joint5", "joint6"};
        arm_joint_order_ = loadArmJointOrder();

        subscription_ = this->create_subscription<JointState>(
            "/joint_states", 10,
            std::bind(&SaveArmPose::jointStateCallback, this, std::placeholders::_1));
        publisher_ = this->create_publisher<JointState>("/saved_arm_pose", 10);
    }

private:
    // ── helpers ──────────────────────────────────────────────────────────

    std::string readFile(const std::string & path)
    {
        std::ifstream ifs(path);
        if (!ifs.is_open()) return {};
        std::ostringstream ss;
        ss << ifs.rdbuf();
        return ss.str();
    }

    bool writeFile(const std::string & path, const std::string & content)
    {
        std::ofstream ofs(path);
        if (!ofs.is_open()) return false;
        ofs << content;
        return ofs.good();
    }

    // ── load joint order from existing xacro ─────────────────────────────

    std::vector<std::string> loadArmJointOrder()
    {
        if (!std::ifstream(xacro_path_).good()) {
            RCLCPP_WARN(get_logger(), "%s not found. Using default joint order.", xacro_path_.c_str());
            return default_arm_joint_order_;
        }

        std::string content = readFile(xacro_path_);
        if (content.empty()) {
            RCLCPP_WARN(get_logger(), "Could not read %s. Using default joint order.", xacro_path_.c_str());
            return default_arm_joint_order_;
        }

        // Find first <group_state ... group="${prefix}xarm6" ...> ... </group_state>
        std::regex group_re(
            R"(<group_state[^>]*group="\$\{prefix\}xarm6"[^>]*>([\s\S]*?)</group_state>)");
        std::smatch group_match;
        if (!std::regex_search(content, group_match, group_re)) {
            RCLCPP_WARN(get_logger(), "No group_state for xarm6 found. Using default joint order.");
            return default_arm_joint_order_;
        }

        std::string block = group_match[1].str();
        std::regex joint_re(R"(<joint[^>]*name="\$\{prefix\}([^"]+)"[^>]*/?>)");
        std::vector<std::string> names;
        auto it = std::sregex_iterator(block.begin(), block.end(), joint_re);
        for (; it != std::sregex_iterator(); ++it) {
            names.push_back((*it)[1].str());
        }

        if (names.empty()) {
            RCLCPP_WARN(get_logger(), "No joints parsed from group_state. Using default joint order.");
            return default_arm_joint_order_;
        }

        std::ostringstream oss;
        for (size_t i = 0; i < names.size(); ++i) {
            if (i) oss << ", ";
            oss << names[i];
        }
        RCLCPP_INFO(get_logger(), "Detected arm joint order: [%s]", oss.str().c_str());
        return names;
    }

    // ── subscription callback ────────────────────────────────────────────

    void jointStateCallback(const JointState::SharedPtr msg)
    {
        if (done_saving_) return;

        publisher_->publish(*msg);

        if (!save_loop_started_) {
            save_loop_started_ = true;
            RCLCPP_INFO(get_logger(), "Starting pose save session...");
            savePose();
            done_saving_ = true;
            RCLCPP_INFO(get_logger(), "Pose saving complete. Shutting down...");
            rclcpp::shutdown();
        }
    }

    // ── wait for a single JointState message ─────────────────────────────

    JointState::SharedPtr waitForJointState(double timeout_sec)
    {
        JointState::SharedPtr received;
        auto sub = this->create_subscription<JointState>(
            "/joint_states", 10,
            [&received](const JointState::SharedPtr msg) {
                if (!received) received = msg;
            });

        auto start = std::chrono::steady_clock::now();
        while (!received && rclcpp::ok()) {
            rclcpp::spin_some(this->shared_from_this());
            auto elapsed = std::chrono::steady_clock::now() - start;
            if (std::chrono::duration<double>(elapsed).count() >= timeout_sec) break;
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }
        return received;
    }

    // ── find a joint value (exact or suffix match) ───────────────────────

    bool findJointValue(
        const std::map<std::string, double> & joint_map,
        const std::string & joint_name,
        double & out)
    {
        auto it = joint_map.find(joint_name);
        if (it != joint_map.end()) { out = it->second; return true; }

        std::vector<double> suffix_matches;
        for (const auto & [name, value] : joint_map) {
            if (name.size() >= joint_name.size() &&
                name.compare(name.size() - joint_name.size(), joint_name.size(), joint_name) == 0)
            {
                suffix_matches.push_back(value);
            }
        }
        if (suffix_matches.size() == 1) { out = suffix_matches[0]; return true; }
        return false;
    }

    // ── interactive save loop ────────────────────────────────────────────

    void savePose()
    {
        std::regex name_re(R"(^[A-Za-z0-9_.\-]+$)");

        while (rclcpp::ok()) {
            std::cout << "Move the arm to the desired pose and enter its name "
                         "(e.g., 'PrePickup', 'LookToGarbage'): " << std::flush;
            std::string arm_name;
            if (!std::getline(std::cin, arm_name)) break;

            // trim
            arm_name.erase(0, arm_name.find_first_not_of(" \t\r\n"));
            arm_name.erase(arm_name.find_last_not_of(" \t\r\n") + 1);

            if (arm_name.empty()) {
                RCLCPP_WARN(get_logger(), "Pose name cannot be empty. Pose not saved.");
                continue;
            }
            if (!std::regex_match(arm_name, name_re)) {
                RCLCPP_WARN(get_logger(),
                    "Invalid pose name. Use only letters, numbers, '-', '_' or '.'. Pose not saved.");
                continue;
            }

            auto msg = waitForJointState(2.0);
            if (!msg) {
                RCLCPP_WARN(get_logger(), "Timed out waiting for /joint_states. Pose not saved.");
                continue;
            }

            std::map<std::string, double> joint_map;
            for (size_t i = 0; i < msg->name.size(); ++i) {
                joint_map[msg->name[i]] = msg->position[i];
            }

            std::vector<double> ordered_pose;
            std::vector<std::string> missing;
            for (const auto & jn : arm_joint_order_) {
                double val;
                if (findJointValue(joint_map, jn, val)) {
                    ordered_pose.push_back(val);
                } else {
                    missing.push_back(jn);
                }
            }

            if (!missing.empty()) {
                std::ostringstream oss;
                for (size_t i = 0; i < missing.size(); ++i) {
                    if (i) oss << ", ";
                    oss << missing[i];
                }
                RCLCPP_WARN(get_logger(),
                    "Missing joints in /joint_states: [%s]. Pose not saved.", oss.str().c_str());
                continue;
            }

            if (writeToXacro(arm_name, ordered_pose)) {
                RCLCPP_INFO(get_logger(), "Saved '%s' to %s", arm_name.c_str(), xacro_path_.c_str());
            }

            std::cout << "Do you want to add more poses? (y/n): " << std::flush;
            std::string answer;
            if (!std::getline(std::cin, answer)) break;
            answer.erase(0, answer.find_first_not_of(" \t\r\n"));
            answer.erase(answer.find_last_not_of(" \t\r\n") + 1);
            std::transform(answer.begin(), answer.end(), answer.begin(), ::tolower);

            if (answer == "n" || answer == "no") {
                break;
            } else if (answer == "y" || answer == "yes") {
                continue;
            } else {
                RCLCPP_WARN(get_logger(), "Invalid input. Please enter 'y' or 'n'. Continuing by default.");
            }
        }
    }

    // ── format a <group_state> block ─────────────────────────────────────

    std::string formatGroupState(const std::string & pose_name,
                                 const std::vector<double> & pose_values)
    {
        std::ostringstream oss;
        oss << "    <group_state name=\"" << pose_name
            << "\" group=\"" << group_name_ << "\">\n";

        for (size_t i = 0; i < arm_joint_order_.size(); ++i) {
            // Use full precision like the Python ".17g"
            char buf[64];
            std::snprintf(buf, sizeof(buf), "%.17g", pose_values[i]);
            oss << "      <joint name=\"${prefix}" << arm_joint_order_[i]
                << "\" value=\"" << buf << "\" />\n";
        }
        oss << "    </group_state>";
        return oss.str();
    }

    // ── write / update the xacro file ────────────────────────────────────

    bool writeToXacro(const std::string & pose_name,
                      const std::vector<double> & pose_values)
    {
        if (!std::ifstream(xacro_path_).good()) {
            RCLCPP_ERROR(get_logger(), "%s does not exist. Pose not saved.", xacro_path_.c_str());
            return false;
        }

        std::string content = readFile(xacro_path_);
        if (content.empty()) {
            RCLCPP_ERROR(get_logger(), "Error reading %s.", xacro_path_.c_str());
            return false;
        }

        std::string block = formatGroupState(pose_name, pose_values);

        // Try to replace an existing group_state with the same name
        std::string escaped_name = std::regex_replace(pose_name, std::regex(R"([.^$|()\\*+?{}\[\]])"), R"(\$&)");
        std::regex existing_re(
            R"([ \t]*<group_state\s+name=")" + escaped_name +
            R"("\s+group="\$\{prefix\}xarm6">[\s\S]*?</group_state>\s*\n?)");

        std::string updated;
        if (std::regex_search(content, existing_re)) {
            updated = std::regex_replace(content, existing_re, block + "\n",
                                         std::regex_constants::format_first_only);
        } else {
            // Insert before <!-- gripper --> comment
            std::regex gripper_re(R"([ \t]*<!--\s*gripper\s*-->)");
            std::smatch gm;
            if (std::regex_search(content, gm, gripper_re)) {
                auto pos = static_cast<size_t>(gm.position(0));
                updated = content.substr(0, pos) + block + "\n\n" + content.substr(pos);
            } else {
                // Fallback: insert before </xacro:macro>
                std::string close_tag = "  </xacro:macro>";
                auto pos = content.rfind(close_tag);
                if (pos == std::string::npos) {
                    RCLCPP_ERROR(get_logger(), "Could not find insertion point in xacro file.");
                    return false;
                }
                updated = content.substr(0, pos) + block + "\n" + content.substr(pos);
            }
        }

        if (!writeFile(xacro_path_, updated)) {
            RCLCPP_ERROR(get_logger(), "Error writing %s.", xacro_path_.c_str());
            return false;
        }
        return true;
    }

    // ── members ──────────────────────────────────────────────────────────

    std::string xacro_path_;
    std::string group_name_;
    std::vector<std::string> default_arm_joint_order_;
    std::vector<std::string> arm_joint_order_;

    rclcpp::Subscription<JointState>::SharedPtr subscription_;
    rclcpp::Publisher<JointState>::SharedPtr publisher_;

    bool save_loop_started_ = false;
    bool done_saving_ = false;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SaveArmPose>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: fbot_manipulator/src/utils/save_arm_pose.py
import os
import re

import rclpy
from rclpy.node import Node
from rclpy.wait_for_message import wait_for_message

from sensor_msgs.msg import JointState

class SaveArmPose(Node):
    def __init__(self):
        super().__init__("save_arm_pose")

        self.declare_parameter(
            "xacro_path",
            "/home/insider/fbot_ws/src/xarm_ros2/xarm_moveit_config/srdf/_xarm6_macro.srdf.xacro",
        )
        self.xacro_path = self.get_parameter("xacro_path").get_parameter_value().string_value

        self.subscription_ = self.create_subscription(JointState, "/joint_states", self.joint_state_callback, 10)
        self.publisher_ = self.create_publisher(JointState, "/saved_arm_pose", 10)
        self.save_loop_started = False
        self.done_saving = False

        self.group_name = "${prefix}xarm6"
        self.default_arm_joint_order = ["joint1", "joint2", "joint3", "joint4", "joint5", "joint6"]
        self.arm_joint_order = self.loadArmJointOrder()

    def loadArmJointOrder(self):
        if not os.path.exists(self.xacro_path):
            self.get_logger().warning(
                f"{self.xacro_path} not found. Using default order {self.default_arm_joint_order}."
            )
            return list(self.default_arm_joint_order)

        try:
            with open(self.xacro_path, "r", encoding="utf-8") as xacro_file:
                content = xacro_file.read()
        except OSError as exc:
            self.get_logger().warning(
                f"Could not read {self.xacro_path}: {exc}. Using default joint order."
            )
            return list(self.default_arm_joint_order)

        group_state_match = re.search(
            r'<group_state[^>]*group="\$\{prefix\}xarm6"[^>]*>(.*?)</group_state>',
            content,
            re.DOTALL,
        )
        if not group_state_match:
            self.get_logger().warning(
                f"Could not infer arm joint order from {self.xacro_path}. Using default order {self.default_arm_joint_order}."
            )
            return list(self.default_arm_joint_order)

        joint_names = re.findall(
            r'<joint[^>]*name="\$\{prefix\}([^"]+)"[^>]*/?>',
            group_state_match.group(1),
        )
        if not joint_names:
            self.get_logger().warning(
                f"No arm joints found in existing group_state. Using default order {self.default_arm_joint_order}."
            )
            return list(self.default_arm_joint_order)

        self.get_logger().info(f"Detected arm joint order: {joint_names}")
        return joint_names
    
    
    def joint_state_callback(self, msg):
        if self.done_saving:
            return
            
        self.publisher_.publish(msg)

        if not self.save_loop_started:
            self.save_loop_started = True
            self.get_logger().info("Starting pose save session...")
            self.savePose()
            self.done_saving = True
            self.get_logger().info("Pose saving complete. Shutting down...")
            rclpy.shutdown()

    def savePose(self) -> None:
        while rclpy.ok():
            arm_name = input("Move the arm to the desired pose and enter its name (e.g., 'PrePickup', 'LookToGarbage'): ").strip()
            if not arm_name:
                self.get_logger().warning("Pose name cannot be empty. Pose not saved.")
                continue

            if not re.fullmatch(r"[A-Za-z0-9_.-]+", arm_name):
                self.get_logger().warning(
                    "Invalid pose name. Use only letters, numbers, '-', '_' or '.'. Pose not saved."
                )
                continue

            success, message = wait_for_message(msg_type= JointState, node=self, topic='/joint_states', time_to_wait=2)
            if not success or message is None:
                self.get_logger().warning("Timed out waiting for /joint_states. Pose not saved.")
                continue

            joints = message.name
            values = message.position
            joint_to_value = dict(zip(joints, values))

            ordered_pose = []
            missing_joints = []
            for joint_name in self.arm_joint_order:
                joint_value = self.findJointValue(joint_to_value, joint_name)
                if joint_value is None:
                    missing_joints.append(joint_name)
                    continue
                ordered_pose.append(joint_value)

            if missing_joints:
                self.get_logger().warning(f"Missing joints in /joint_states: {missing_joints}. Pose not saved.")
                continue

            if self.writeToXacro(arm_name, ordered_pose):
                self.get_logger().info(f"Saved '{arm_name}' to {self.xacro_path}")

            save_more = input("Do you want to add more poses? (y/n): ").strip().lower()
            if save_more in ('n', 'no'):
                break
            elif save_more in ('y', 'yes'):
                continue
            else:
                self.get_logger().warning("Invalid input. Please enter 'y' or 'n'. Continuing by default.")
                continue

    def findJointValue(self, joint_to_value, joint_name):
        if joint_name in joint_to_value:
            return float(joint_to_value[joint_name])

        suffix_matches = [
            float(value)
            for name, value in joint_to_value.items()
            if name.endswith(joint_name)
        ]
        if len(suffix_matches) == 1:
            return suffix_matches[0]

        return None

    def formatGroupState(self, pose_name, pose_values):
        lines = [f'    <group_state name="{pose_name}" group="{self.group_name}">']
        for joint_name, joint_value in zip(self.arm_joint_order, pose_values):
            value_as_text = format(float(joint_value), ".17g")
            lines.append(
                f'      <joint name="${{prefix}}{joint_name}" value="{value_as_text}" />'
            )
        lines.append("    </group_state>")
        return "\n".join(lines)

    def writeToXacro(self, pose_name, pose_values):
        if not os.path.exists(self.xacro_path):
            self.get_logger().error(f"{self.xacro_path} does not exist. Pose not saved.")
            return False

        try:
            with open(self.xacro_path, "r", encoding="utf-8") as xacro_file:
                content = xacro_file.read()
        except OSError as exc:
            self.get_logger().error(f"Error reading {self.xacro_path}: {exc}")
            return False

        group_state_block = self.formatGroupState(pose_name, pose_values)

        existing_pose_pattern = re.compile(
            rf'^[ \t]*<group_state\s+name="{re.escape(pose_name)}"\s+group="\$\{{prefix\}}xarm6">.*?^[ \t]*</group_state>\s*\n?',
            re.MULTILINE | re.DOTALL,
        )

        if existing_pose_pattern.search(content):
            updated_content = existing_pose_pattern.sub(group_state_block + "\n", content, count=1)
        else:
            insertion_match = re.search(r"^[ \t]*<!--\s*gripper\s*-->.*$", content, re.MULTILINE)
            if insertion_match:
                insert_at = insertion_match.start()
                updated_content = content[:insert_at] + group_state_block + "\n\n" + content[insert_at:]
            else:
                macro_close_tag = "  </xacro:macro>"
                close_index = content.rfind(macro_close_tag)
                if close_index == -1:
                    self.get_logger().error("Could not find where to insert new <group_state> block.")
                    return False
                updated_content = content[:close_index] + group_state_block + "\n" + content[close_index:]

        try:
            with open(self.xacro_path, "w", encoding="utf-8") as xacro_file:
                xacro_file.write(updated_content)
        except OSError as exc:
            self.get_logger().error(f"Error writing {self.xacro_path}: {exc}")
            return False

        return True
       

       
def main(args=None):
    rclpy.init(args=args)
    node = SaveArmPose()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
if __name__ == "__main__":
    main()